src/cassette.jl

```diff
@@ -141,6 +143,143 @@ function replacefunc(f::Function, d::AbstractDict)
     ctx = GraftCtx(metadata=d)
     return Cassette.recurse(ctx, f)
 end
+            
+#------------------------------------------------------------------------------------           
+            
+Cassette.@context TypeCtx
+            
+"""   TypeCtx
+
+creates a MetaDiGraph tracking the types of args and ret values throughout a script
+
+"""
+TypeCtx
+            
+            
+
+"""   FCollector(depth::Int,frame::function,data::FCollector)
+
+struct to collect all the functions called throughout a script
+        
+"""
+mutable struct FCollector{I,F,C}
+    depth::I
+    frame::F
+    data::Vector{C}
+end
+            
+
+
+"""   FCollector(depth::Int,frame::Frame)
+
+this is an initialization funtion for the FCollector
+
+"""
+function FCollector(d::Int, f)
+    FCollector(d, f, FCollector[])
+end
+
+""" Frame(func, args, ret, subtrace)
+
+a structure to hold metadata for recursive type information for every function called
+            
+"""
+mutable struct Frame{F,T,U}
+    func::F
+    args::T
+    ret::U
+end
+            
+# add boilerplate for functionality
+function Cassette.overdub(ctx::TypeCtx, f, args...)
+    c = FCollector(ctx.metadata.depth-1, Frame(f, args, Any))
+    push!(ctx.metadata.data, c)
+    if c.depth > 0 && Cassette.canrecurse(ctx, f, args...)
+        newctx = Cassette.similarcontext(ctx, metadata = c)
+        z = Cassette.recurse(newctx, f, args...)
+        c.frame.ret = typeof(z)
+        return z
+    else
+        z = Cassette.fallback(ctx, f, args...)
+        c.frame.ret = typeof(z)
+        return z
+    end
+end
+
+# limit the stacktrace in terms of which to recurse on
+Cassette.canrecurse(ctx::TypeCtx,::typeof(Base.vect), args...) = false
+Cassette.canrecurse(ctx::TypeCtx,::typeof(FCollector)) = false
+Cassette.canrecurse(ctx::TypeCtx,::typeof(Frame)) = false
+
+# this is a function that strips all using statements from a raw text file
+# reasons to do so
+# 1) we no longer include libraries we dont have too so semanticmodels can remain small
+# 2) prevents errors in the include-cassette design which doesn allow for usings
+function extractdeps(filename::String)
+    f = open(filename)
+    newf = open("new$filename","w") 
+    for line in readlines(f)
+        if occursin("using",line)
+            ex = Meta.parse(line)
+            eval(ex)
+        else
+            write(newf,"$line\n")
+        end
+    end
+    close(f)
+    close(newf)
+    return "new$filename"
+end
+                
+""" buildgraph
+
+internal function used in the typegraphfrompath
+takes the collector object and returns a metagraph
+            
+"""
+function buildgraph(g,collector)
+    add_vertex!(g,:name,collector.frame.args)
```

Comment:
You need to check if `collector.frame.args` already exists in the graph and only `add_vertex!` if it does not already exist. I think this code will just create a path graph or bunch of isolated edges, because every edge that gets inserted is between `nv(g)` and `nv(g)-1` which are being incremented on each recursive call.

Reply:
MetaGraphs docs aren't clear to me, but it seems as though an indexing prop can only be set after the full graph has been constructed?
![image](https://user-images.githubusercontent.com/41996797/53341293-2ddc9400-38d9-11e9-8295-f55d9b5bd800.png)

Reply:
you can use `set_indexing_prop!(G, 3, :name, "name3")` to update the index after you construct the index with `set_indexing_prop!(G, :name)`